src/pudl/analysis/demand_mapping.py

```diff
@@ -1,294 +1,1009 @@
+    the US Census DP1 dataset with each record, based on the county FIPS ID. This
+    (enormous) GeoDataFrame can then be used to produce simpler annual geometries by
+    dissolving based on either balancing authority or utility IDs and the report date.
+
+    The way that the relationship between balancing authorities and utilities is
+    reported changed between 2012 and 2013. Prior to 2013, the EIA 861 balancing
+    authority table enumerates all of the utilities which participate in each balancing
+    authority. In 2013 and subsequent years, the balancing authority table associates a
+    balancing authority code (e.g. SWPP or ERCO) with each balancing authority ID, and
+    also lists which states the balancing authority was operating in. These balancing
+    authority codes then appear in other EIA 861 tables like the Sales table, in
+    association with utilities and often states. For these later years, we must compile
+    the list of utility IDs which are seen in association with a particular balancing
+    authority code to understand which utilities are operating within which balancing
+    authorities, and thus which counties should be included in that authority's
+    territory. Because the state is also listed, we can select only a subset of the
+    counties that are part of the utility, providing much more geographic specificity.
+    This is especially important in the case of sprawling western utilities like
+    PacifiCorp, which drastically expand the apparent territory of a balancing authority
+    if the utility's entire service territory is included just because the sold
+    electricty within one small portion of the balancing authority's territory.
+
+    Args:
+        ba_eia861 (pandas.DataFrame): The balancing_authority_eia861 table.
+        st_eia861 (pandas.DataFrame): The service_territory_eia861 table.
+        sales_eia861 (pandas.DataFrame): The sales_eia861 table.
+        census_gdf (geopandas.GeoDataFrame): The counties layer of the US Census DP1
+            geospatial dataset.
+        output_crs (str): String representing a coordinate reference system (CRS) that
+            is recognized by geopandas. Applied to the output GeoDataFrame.
+
+    Returns:
+        geopandas.GeoDataFrame: Contains columns identifying the balancing authority,
+        utility, and state, along with the county geometry, for each year in which
+        those balancing authorities / utilities appeared in the EIA 861 Balancing
+        Authority table (through 2012) or the EIA 861 Sales table (for 2013 onward).
+
     """
-    Re-allocate demand in a time series from source to target geometries.
+    # Make sure that there aren't any more BA IDs we can recover from later years:
+    ba_ids_missing_codes = (
+        ba_eia861.loc[
+            ba_eia861.balancing_authority_code_eia.isnull(),
+            "balancing_authority_id_eia"]
+        .drop_duplicates()
+        .dropna()
+    )
+    assert len(ba_eia861[
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/zaneselvans/pudl/pullRequest?prid=5664147)